RCTImage: Use C atomics instead of OSAtomic

Adlai-Holler opened the pull request:
The next in my series of :atom: migrations.

Adlai-Holler commented on Libraries/Image/RCTImageLoader.m:
```diff
@@ -524,20 +528,25 @@ - (RCTImageLoaderCancellationBlock)loadImageWithURLRequest:(NSURLRequest *)image
+        }
         dispatch_block_t cancelLoadLocal = cancelLoad;
-        if (cancelLoadLocal && !cancelled) {
+        cancelLoad = nil;
```
I added this line because it seems to have been missing from previous implementation. The pattern of "read cancel block into local variable, erase the __block variable, run the local variable" was copied from line 400 but it was missing this line, which defeats the purpose.

shergin commented:
@Adlai-Holler That looks great! But... Could you please clarify the motivation? How `stdatomic` is better than `osatomic`? Which problem do we solve?

Adlai-Holler commented:
Sure! We get a few benefits:

- OSAtomic is deprecated, so we get off from that: https://opensource.apple.com/source/libplatform/libplatform-125/include/libkern/OSAtomicDeprecated.h.auto.html https://github.com/google/protobuf/issues/2182 https://forums.developer.apple.com/thread/62765
- We remove function calls – stdatomic is fully inlined and ends up as one or two instructions.
- The API is more purpose-built and flexible: `atomic_bool` vs `volatile int32_t`.

It doesn't exactly shatter the Earth but I noticed it when I started exploring the code base.

Sent from my iPhone

> On Jul 30, 2017, at 3:29 PM, Valentin Shergin <notifications@github.com> wrote:
> 
> @Adlai-Holler That looks great! But... Could you please clarify the motivation? How stdatomic is better than osatomic? Which problem do we solve?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

javache commented:
Actually, it would be fine to make these Objective-C++ as well, and just use std::atomic everywhere.

javache commented on Libraries/Image/RCTImageStoreManager.m:
```diff
@@ -138,13 +138,14 @@ - (BOOL)canHandleRequest:(NSURLRequest *)request
 - (id)sendRequest:(NSURLRequest *)request withDelegate:(id<RCTURLRequestDelegate>)delegate
 {
   __block volatile uint32_t cancelled = 0;
```
You forgot to remove the old definition here

Adlai-Holler commented:
@javache I've rebased and removed the leftover def. I'm willing to migrate these to C++ but it'll be easier if there's an xcodeproj I can use that will build the files. If I use `React.xcodeproj` the RCTImage implementation files aren't included, and the `ReactLegacy.xcodeproj` and `RCTImage.xcodeproj` don't seem to want to build on their own. Any advice for a solid dev environment?

javache commented:
I typically use the RNTester.xcodeproj (https://github.com/facebook/react-native/tree/master/RNTester/RNTester.xcodeproj)
